src/queues/mod.rs

```diff
@@ -0,0 +1,209 @@
+
+use std::{
+    boxed::Box, error::Error, fmt::{self, Debug, Display, Formatter},
+};
+
+use futures::future::{self, Future};
+
+use self::notification::{Notification, NotificationType};
+pub use self::sqs::Queue as Sqs;
+use auth_db::{BounceSubtype, BounceType, Db, DbClient, DbError};
+use settings::Settings;
+
+mod mock;
+mod notification;
+pub mod sqs;
+#[cfg(test)]
+mod test;
+
+#[derive(Debug)]
+pub struct Queues {
+    bounce: Box<Incoming>,
+    complaint: Box<Incoming>,
+    delivery: Box<Incoming>,
+    notification: Box<Outgoing>,
+    db: DbClient,
+}
+
+pub trait Incoming: Debug + Sync {
+    fn receive(&'static self) -> ReceiveFuture;
+    fn delete(&'static self, message: Message) -> DeleteFuture;
+}
+
+type ReceiveFuture = Box<Future<Item = Vec<Message>, Error = QueueError>>;
+type DeleteFuture = Box<Future<Item = (), Error = QueueError>>;
+
+pub trait Outgoing: Debug + Sync {
+    fn send(&'static self, body: &Notification) -> SendFuture;
+}
+
+type SendFuture = Box<Future<Item = String, Error = QueueError>>;
+
+pub trait Factory {
+    fn new(id: String, settings: &Settings) -> Self;
+}
+
+#[derive(Debug, Default)]
+pub struct Message {
+    pub id: String,
+    pub notification: Notification,
+}
+
+#[derive(Debug)]
+pub struct QueueError {
+    description: String,
+}
+
+#[derive(Debug)]
+pub struct QueueIds {
+    pub bounce: String,
+    pub complaint: String,
+    pub delivery: String,
+    pub notification: String,
+}
+
+impl Queues {
+    pub fn new<Q: 'static>(ids: QueueIds, settings: &Settings) -> Queues
+    where
+        Q: Incoming + Outgoing + Factory,
+    {
+        Queues {
+            bounce: Box::new(Q::new(ids.bounce, settings)),
+            complaint: Box::new(Q::new(ids.complaint, settings)),
+            delivery: Box::new(Q::new(ids.delivery, settings)),
+            notification: Box::new(Q::new(ids.notification, settings)),
+            db: DbClient::new(settings),
+        }
+    }
+
+    pub fn process(&'static self) -> QueueFuture {
+        let joined_futures = self
+            .process_queue(&self.bounce)
+            .join3(
+                self.process_queue(&self.complaint),
+                self.process_queue(&self.delivery),
+            )
+            .map(|results| results.0 + results.1 + results.2);
+
+        Box::new(joined_futures)
+    }
+
+    fn process_queue(&'static self, queue: &'static Box<Incoming>) -> QueueFuture {
+        let future = queue
+            .receive()
+            .and_then(move |messages| {
+                let mut futures: Vec<Box<Future<Item = (), Error = QueueError>>> = Vec::new();
+                for message in messages.into_iter() {
+                    if message.notification.notification_type != NotificationType::Null {
+                        let future = self
+                            .handle_notification(&message.notification)
+                            .and_then(move |_| queue.delete(message));
```

Comment:
Since we receive messages in batches, I wonder if we should also delete them in batches using [1], or whether that would require us to jump through some ugly hoops in the processing loop here.

[1] https://rusoto.github.io/rusoto/rusoto_sqs/trait.Sqs.html#tymethod.delete_message_batch

Reply:
Oh, good shout. :+1:

Reply:
(I opted to leave this alone for now, but have opened #54 to so we can come back to it)